### PR TITLE
Made to compile by MSVC 2015/2017 32 bit compilers

### PR DESCRIPTION
**Pull request by anatoly-spb**
Added several fixes to compile by MSVC 2015/2017 32 bit compilers

### CODE REVIEW

**Comment by crackedmind** [comment]
:x: [Build kepka 1.0.71-epouxvof failed](https://ci.appveyor.com/project/procxx/tdesktop/build/1.0.71-epouxvof) (commit https://github.com/procxx/kepka/commit/eda31b9d95 by @anatoly-spb)

File: Telegram/SourceFiles/base/assertion.h [hunk 1]
```diff
@@ -43,10 +43,15 @@ inline constexpr void noop() {
 	std::abort();
 }
 
-inline constexpr void validate(bool condition, const char *message, const char *file, int line) {
+inline
+#if !defined(_MSC_VER) || (_MSC_VER > 1900)
+constexpr
+#endif
+  void validate(bool condition, const char *message, const char *file, int line) {
```
Comment (ForNeVeR) [inline Telegram/SourceFiles/base/assertion.h:52]: I don't get what problem we're trying to solve with this change. Some version of MSVC doesn't support `constexpr void`? I suggest to drop support for that compiler; we do require a decent support of C++17.

File: Telegram/SourceFiles/core/basic_types.h [hunk 2]
```diff
@@ -54,4 +54,4 @@ void as_const(const T&&) = delete;
 #include <cstdint>
 
 #define qsl(s) QStringLiteral(s)
-#define qstr(s) QLatin1String(s, sizeof(s) - 1)
+#define qstr(s) QLatin1String(s, s + sizeof(s))
```
Comment (ForNeVeR) [inline Telegram/SourceFiles/core/basic_types.h]: Looks like this change caused [a build failure](https://ci.appveyor.com/project/procxx/tdesktop/build/1.0.71-epouxvof#L505) (although I don't understand why); what's the purpose of it?

Comment (ForNeVeR) [reply Telegram/SourceFiles/core/basic_types.h]: It looks like the [corresponding signature for `QLatin1String` constructor](http://doc.qt.io/qt-5/qlatin1string.html#QLatin1String-2) was introduced in Qt 5.10, although we claim to support Qt 5.9+. [Our build server uses Qt 5.9](https://github.com/procxx/kepka/blob/2a2d4a260ec1c91aa73a0ccc670f5f5ca18fcd3c/appveyor.yml#L26), and that's likely the reason for the build failure.

Do you suggest us to migrate to Qt 5.10? If so, then please clearly state the reason.

File: Telegram/SourceFiles/base/assertion.h (hunk 1 above)
Comment (anatoly-spb) [reply Telegram/SourceFiles/base/assertion.h:52]: It's only one peace of code which is not compiled by MSVC 2015 Update 3:
```
inline constexpr void validate(bool condition, const char *message, const char *file, int line) {
(47)  (GSL_UNLIKELY(!(condition))) ? fail(message, file, line) : noop();
}
```
3>h:\tmp\kepka\telegram\sourcefiles\base/assertion.h(47): error C3249: illegal statement or sub-expression for 'constexpr' function

So we have decided to  use conditional compilation for only this function.

File: Telegram/SourceFiles/core/basic_types.h (hunk 2 above)
Comment (anatoly-spb) [reply Telegram/SourceFiles/core/basic_types.h]: It's my fault, it was elegant solution but it is compiled only with Qt 5.10 and newer. I plan to revert to old version.

**Comment by crackedmind** [comment]
:white_check_mark: [Build kepka 1.0.71-yorqedks completed](https://ci.appveyor.com/project/procxx/tdesktop/build/1.0.71-yorqedks) (commit https://github.com/procxx/kepka/commit/36aa1951cf by @anatoly-spb)

**Comment by crackedmind** [comment]
:white_check_mark: [Build kepka 1.0.72-qlbvtqps completed](https://ci.appveyor.com/project/procxx/tdesktop/build/1.0.72-qlbvtqps) (commit https://github.com/procxx/kepka/commit/8481615d84 by @anatoly-spb)

**Pull request merged**